### PR TITLE
Fix #10338: Translcore.push_defaults does not respect scoping

### LINKED ISSUE #10338: Translcore.push_defaults does not respect scoping
Evaluation of defaults of optional parameters is pushed under function binders in a way that is not always sound:

```
        OCaml version 4.12.0

# let s = 42;;
val s : int = 42
# let f ?(s="hello") = function x when (print_string s; true) -> x;;
Warning 8 [partial-match]: this pattern-matching is not exhaustive.
All clauses in this pattern-matching are guarded.
val f : ?s:string -> 'a -> 'a = <fun>
# f ();;
Segmentation fault
```

`f` gets translated to something resembling:
```ocaml
let f opt = function
  | x when (print_string s; true) ->
    let s = match opt with Some s -> s | None -> "hello" in
    x
```
where the `s` in the guard is the wrong one.

### PR DESCRIPTION
**Pull request by garrigue**
Strangely, the check `c_guard=None` was missing in just one case of `Translcore.push_defaults`.
The role of the next case is precisely to build a case where `c_guard=None`, so let's just defer to it.

### CODE REVIEW

**Comment by stedolan** [comment]
Thanks for the quick fix, it's simpler than I expected.

However, I'm still suspicious of the reordering that occurs here. Some more worries around `push_defaults`:

1. ~~guards can refer to values bound by optional parameters~~ - fixed here

2. side-effects can be reordered:
      `let f ?(y = assert false) (lazy e) = 1 in f (lazy (print_endline "hello"))`
      prints "hello", which surprises me.

3. typing of later patterns can depend on earlier ones:
      `let f (type a) ?c:(Refl = (assert false : (a, (int * int)) eq)) ((p, q) : a) = p + q` is strange.
      (This doesn't seem to crash, I think because the GADT match generates a Texp_match rather than a Texp_let? I don't really understand what's going on here)

4. what happens when there are no cases? The 4.12 behaviour is strange:

    ```  
            OCaml version 4.12.0

    # type empty = |;;
    type empty = |
    # let f : empty -> int = function _ -> .;;
    val f : empty -> int = <fun>
    # let f ?y : empty -> int = function _ -> .;;
    val f : ?y:'a -> empty -> int = <fun>
    # let f ?(y=1) : empty -> int = function _ -> .;;
    Error: Unreachable expression was reached
    ```

5. the pushing of module bindings is also suspect, and does not require guards to trigger bad behaviour:

    ```
            OCaml version 4.12.0

    # module type E = sig exception Ex end;;
    module type E = sig exception Ex end
    # let f ((module M) : (module E)) (M.Ex | _) = 42;;
    >> Fatal error: M unbound at toplevel
    Exception: Misc.Fatal_error.
    ```

**Comment by garrigue** [comment]
Indeed, all the cases you present are problematic.
While we allow some changes in evaluation order, for instance delaying of evaluation of partially applied functions, we still expect side-effects in patterns to occur in the expected order.

This code is there to avoid blow-ups in compilation due to functions taking a lot of arguments, and interspersing `Texp_function` and `Texp_let`. If we were to respect strictly the evaluation order, we would have sometimes to generate all the possible combinations of application lengths, which can kill ocamlopt.
However it was written a long time ago, when we still believed that patterns were nice declarative things.

I have now restricted this code to work only on trivial cases, i.e. where the patterns are trivial ones (variable or unit, basically), there is no guard, and the rhs is not `Texp_unreachable`. This may degrade the compilation is some more complex cases, but this falls outside of its scope.

**Comment by gasche** [comment]
(Do we now have testsuite coverage for the examples Stephen produced?)

**Comment by stedolan** [comment]
Thanks for the explanation. I understand the desire to delay evaluation of optional arguments, but the design here still seems inconsistent to me.

First, what is the relationship between this mechanism and the other mechanism for delaying evaluation of arguments in the presence of lazy/mutable matches? (see `Parmatch.inactive`, #1308, warning 68, #9751).

Second, it fundamentally seems like a mistake to me to reorder evaluation of patterns in arguments. The `push_defaults` logic translates a function definition `let f ?(P = E) Q -> ...` into the following, for patterns P, Q and expression E:
```ocaml
  let f opt arg =
    let Q = arg in
    let P = match opt with Some x -> x | None -> E in
    ...
```
There are bugs in the current implementation involving `Q` being evaluated in the wrong context. I think that the present patch does fix these, by identifying a class of patterns `Q` for which the reordering doesn't break. But surely it would be simpler not to reorder, translating the function instead into:
```ocaml
  let f opt arg =
    let P = match opt with Some x -> x | None -> E in
    let Q = arg in
    ...
```
(My suspicion is that the two points are related: if push_defaults were changed not to reorder pattern evaluation, then I think it might be correct to use the same Parmatch.inactive check for how far they should be pushed. That's not correct under the present reordering though, as it is unsound if e.g. P is a partial GADT match)

**Comment by garrigue** [comment]
@gasche Yes, they are all in pr10338.ml

@stedolan Of course I thought of doing so, and started writing the code. This amounts to adding a `Bind_pattern` case to the bindings. But the trouble is that `push_defaults` runs before `transl_curried_function`, which does some optimizations of its own. So I'm afraid of breaking something by doing too much. And if we do things too eagerly, we may also change behaviors in more cases than intended.

As for the relation to other problems in patterns, the goal of this PR is to restrict to a class of patterns that do not suffer from any of these problems, so we are certainly fine.

As this is just an optimization, the restriction seems fine to me.
I can also introduce `Bind_pattern` as suggested above after introducing the first default/module.
Or even do it systematically if we agree that all pattern-related side effects can always postponed to after accepting all arguments.

Finally, a better solution could be to do this inside `transl_curried_function`, rather than as a pre-processing pass, but this requires care there too.

File: lambda/translcore.ml [hunk 1]
```diff
@@ -88,42 +88,52 @@ type binding =
...
+  | _ -> false
+
+let rec push_defaults loc bindings cases partial need_pat =
   match cases with
     [{c_lhs=pat; c_guard=None;
       c_rhs={exp_desc = Texp_function { arg_label; param; cases; partial; } }
-        as exp}] ->
-      let cases = push_defaults exp.exp_loc bindings cases partial in
+        as exp}] when bindings = [] || trivial_pat pat ->
+      let cases = push_defaults exp.exp_loc bindings cases partial false in
```
Comment (stedolan) [inline lambda/translcore.ml]: I don't understand why need_pat becomes false here

**Comment by stedolan** [comment]
> @gasche Yes, they are all in pr10338.ml
> 
> @stedolan Of course I thought of doing so, and started writing the code. This amounts to adding a `Bind_pattern` case to the bindings. But the trouble is that `push_defaults` runs before `transl_curried_function`, which does some optimizations of its own. So I'm afraid of breaking something by doing too much. And if we do things too eagerly, we may also change behaviors in more cases than intended.
> 
> As for the relation to other problems in patterns, the goal of this PR is to restrict to a class of patterns that do not suffer from any of these problems, so we are certainly fine.

Thanks for the explanation, this makes sense to me.

> As this is just an optimization, the restriction seems fine to me.

I don't think it's exactly an optimisation, since it changes semantics. Most of the examples I've mentioned are strange code, but here is a more reasonable one:

    let f ?(table = Hashtbl.create 20) x y = ...

The sharing of the hashtable depends on exactly how far push_defaults pushes. Since it's user-visible, we should think a bit about which patterns count as trivial (in particular, Tpat_tuple should probably be included)

> I can also introduce `Bind_pattern` as suggested above after introducing the first default/module.
> Or even do it systematically if we agree that all pattern-related side effects can always postponed to after accepting all arguments.
> 
> Finally, a better solution could be to do this inside `transl_curried_function`, rather than as a pre-processing pass, but this requires care there too.

I think your current implementation style makes sense for this PR as it fixes the issue in a minimal way. I'm going to experiment with a larger change, moving the arity-detection to Typedtree (where I think it fits better for several reasons), but that shouldn't delay merging this PR.

**Comment by stedolan** [comment]
(incidentally, the CI failure on i386-static was fixed by #10334, and if you rebase onto current trunk it should go away)

**Comment by gasche** [comment]
@stedolan does this mean that you are Approving the PR?

**Comment by stedolan** [comment]
> @stedolan does this mean that you are Approving the PR?

Nearly? See the comment about need_pat above.

File: lambda/translcore.ml (hunk 1 above)
Comment (garrigue) [reply lambda/translcore.ml]: `need_pat` is true when the `bindings` depend on the `c_lhs` part of cases.
This becomes true when we add something to `bindings` (as it depends on the pattern we have just seen), but becomes false after we switch to new `cases`, as is the case here.
I renamed it to `use_lhs`, but ultimately there should be a better way to do that (somewhere else).

**Comment by garrigue** [comment]
> I don't think it's exactly an optimisation, since it changes semantics. Most of the examples I've mentioned are strange code, but here is a more reasonable one:
> 
> ```
> let f ?(table = Hashtbl.create 20) x y = ...
> ```
> 
> The sharing of the hashtable depends on exactly how far push_defaults pushes. Since it's user-visible, we should think a bit about which patterns count as trivial (in particular, Tpat_tuple should probably be included)

Assuming that the semantics is that the timing of the the evaluation of optional arguments is left undefined, then choosing when we delay it is just an optimization of the code size. This changes the behavior, but inside the same semantics. And your example has just undefined semantics. So, yes this can be observed, but if this changes the behavior of your code, it is actually wrong from the beginning.

Leaving things undefined is not good in general, but I'm not sure what we could specify more here without getting bad compilation behavior for libraries such as labltk and lablgtk.

BTW, I finally made `Tpat_tuple` a trivial pattern. The reason I omitted it originally is that it is hard to decide where to stop. For instance, records are dangerous (mutability), GADTs are even more dangerous, and there seems to be little point in doing a deep analysis here. But there is a point that tuples are a common way to pass arguments.

> I think your current implementation style makes sense for this PR as it fixes the issue in a minimal way. I'm going to experiment with a larger change, moving the arity-detection to Typedtree (where I think it fits better for several reasons), but that shouldn't delay merging this PR.

I'm curious of what you have in mind. I think that this belongs more to backend than frontend, and I would have rather delayed it that advanced it, but I suppose you have your reasons.

**Comment by gasche** [comment]
(I wonder if @stedolan is now completely satisfied with the PR.)

File: lambda/translcore.ml (hunk 1 above)
Comment (stedolan) [reply lambda/translcore.ml]: Thanks for the explanation. It took me a little while to understand this logic, but it makes sense now.

**Comment by stedolan** [comment]
> > I don't think it's exactly an optimisation, since it changes semantics. 

> Assuming that the semantics is that the timing of the the evaluation of optional arguments is left undefined, then choosing when we delay it is just an optimization of the code size. This changes the behavior, but inside the same semantics. And your example has just undefined semantics. So, yes this can be observed, but if this changes the behavior of your code, it is actually wrong from the beginning.

OK, but in the absence of a fully detailed specification I don't really see any distinction between "this optimisation changes semantics" and "the semantics leaves this aspect undefined". Anything can be explained by saying that the semantics is a nondeterministic choice between the program text and what the optimiser does!

> BTW, I finally made Tpat_tuple a trivial pattern. The reason I omitted it originally is that it is hard to decide where to stop. For instance, records are dangerous (mutability), GADTs are even more dangerous, and there seems to be little point in doing a deep analysis here. But there is a point that tuples are a common way to pass arguments.

Thanks. I misread your old `trivial_pat` function: I thought you were counting `Tpat_construct` with a single constructor as "trivial", and it was strange to me to allow `F (a, b)` but not `(a, b)`. I now realise that you were counting as trivial only `Tpat_construct` with a single _constant_ constructor, so that defaults would be pushed past the unit pattern (a very common situation).

(I'm still happy to see `Tpat_construct` here, tupled arguments are common. But that's why I thought this case was particularly important)

> > I'm going to experiment with a larger change, moving the arity-detection to Typedtree (where I think it fits better for several reasons), but that shouldn't delay merging this PR.

> I'm curious of what you have in mind. I think that this belongs more to backend than frontend, and I would have rather delayed it that advanced it, but I suppose you have your reasons.

I have a few:

  - I think that the behaviour of a program should be determined as soon as typing is finished (when scoping / disambiguation / etc. have all been resolved), and that the phases of the compiler beyond typing should be behaviour-preserving transformations. 

    We go to great lengths to enforce this property elsewhere: for instance, despite argument ordering being formally unspecified, the backend is careful to preserve it. Also, the currying optimisations in Lambda are careful not to reorder effects in patterns such as mutable field matching (cf Parmatch.inactive, warning 68, etc.), a situation very analogous to `push_defaults`.

  - I find the current implementation a bit messy and hard to read: during typing, special typedtree nodes are constructed and marked with magic strings (`#default` and `#modulepat` in attributes), so that a later phase can detect them and reorder them. It would be simpler to do this all in one place.

  - More speculatively, I'm planning to experiment with modelling currying in types, rather than discarding arity information in the parser and redetecting it in Lambda. This would entail adding a multi-parameter `Texp_fun` to typedtree (and maybe, one day, Parsetree as well), which then seems like a natural place to do `push_defaults`.

**Review by stedolan — APPROVED** [review]

**Comment by stedolan** [comment]
> (I wonder if @stedolan is now completely satisfied with the PR.)

No, but complete satisfaction is not a precondition for merging! This PR fixes the bug and improves the compiler.

**Pull request merged**